File: app/src/main/java/padev/badhabits/application/ui/activity/HomeActivity.java
package padev.badhabits.application.ui.activity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.facebook.stetho.Stetho;

import java.util.ArrayList;

import butterknife.ButterKnife;
import butterknife.OnClick;
import padev.badhabits.Data.AbstractData;
import padev.badhabits.Data.CRUD.HabitCRUD;
import padev.badhabits.Data.Habit;
import padev.badhabits.R;
import padev.badhabits.application.mvp.presenter.home.HomePresenter;
import padev.badhabits.application.mvp.view.IHomeView;
import padev.badhabits.core.view.BaseActivity;

public class HomeActivity extends BaseActivity implements IHomeView {

    @InjectPresenter
    HomePresenter mHomePresenter;

    private HabitCRUD habitCRUD;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // for SQLite view
        Stetho.initializeWithDefaults(this);

        ButterKnife.bind(this);

        // HabitCRUD is Singleton
        habitCRUD = HabitCRUD.getInstance(this);

        ArrayList<AbstractData> habits =  habitCRUD.selectAllData();

        for (AbstractData data: habits) {

            Habit habit = (Habit) data;
            createHabitCard(habit);
        }
    }

    @OnClick(R.id.activity_main_fab)
    public void addHabit(View view) {

        View promptsView = getLayoutInflater().inflate(R.layout.activity_main_input_alert, null);

        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(this);

        alertDialogBuilder.setView(promptsView);

        final EditText habitName = (EditText) promptsView.findViewById(R.id.habit_name_input_alert_habit_name_edit_text);

        alertDialogBuilder.setCancelable(false);

        alertDialogBuilder.setPositiveButton("Добавить", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

                String hNameStr = habitName.getText().toString();

                Habit habit = new Habit(hNameStr);

                habitCRUD.insertData(habit);

                createHabitCard(habit);
            }
        });

        alertDialogBuilder.setNegativeButton("Отмена", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });

        AlertDialog alertDialog = alertDialogBuilder.create();

        alertDialog.show();
    }

    private void createHabitCard(Habit habit) {

        LinearLayout cardViewLayout = (LinearLayout) getLayoutInflater().inflate(R.layout.activity_main_habit_card, (LinearLayout) findViewById(R.id.activity_main_habits_list), true);

        View currentView = cardViewLayout.getChildAt(cardViewLayout.getChildCount()-1);

        currentView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                habitCardClicked(v);
            }
        });

        TextView habitNameTextView = (TextView) currentView.findViewById(R.id.habit_card_habit_name);

        habitNameTextView.setText(habit.getName());
        habitNameTextView.setTag(habit.getId());
    }

    private void habitCardClicked(View view) {

        TextView habitNameTextView = (TextView) view.findViewById(R.id.habit_card_habit_name);

        Habit habit = new Habit(Integer.parseInt(habitNameTextView.getTag().toString()), habitNameTextView.getText().toString());

        Intent intent = new Intent(this, HabitActivity.class);

        intent.putExtra("habit_name", habit.getName());
        intent.putExtra("habit_id", Long.toString(habit.getId()));

        startActivity(intent);
    }
}

File: app/src/main/java/padev/badhabits/application/ui/activity/HomeActivity.kt
package padev.badhabits.application.ui.activity;

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.Toolbar
import android.view.View
import android.widget.CheckBox
import android.widget.EditText
import android.widget.ProgressBar
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.facebook.stetho.Stetho
import com.mikepenz.iconics.typeface.FontAwesome
import com.mikepenz.materialdrawer.Drawer
import com.mikepenz.materialdrawer.model.DividerDrawerItem
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem
import com.mikepenz.materialdrawer.model.SecondaryDrawerItem
import padev.badhabits.R
import padev.badhabits.application.mvp.model.entity.HabitEntity
import padev.badhabits.application.mvp.presenter.home.HomePresenter
import padev.badhabits.application.mvp.view.IHomeView
import padev.badhabits.application.service.NotificationBuilderService
import padev.badhabits.application.ui.fragment.HabitCardFragment
import padev.badhabits.core.view.BaseActivity

class HomeActivity: BaseActivity(), IHomeView {

    private val TAG = HomeActivity::class.java.simpleName

    private val presenter = HomePresenter(this)

    private lateinit var mDrawerResult: Drawer.Result

    @BindView(R.id.activity_home_fab)
    lateinit var fab: FloatingActionButton

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_home)

        ButterKnife.bind(this)

        val intent = Intent(this, NotificationBuilderService::class.java)
        startService(intent)

        val toolbar = findViewById<Toolbar>(R.id.activity_home_appbar_toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        supportActionBar?.title = getString(R.string.app_name)

        mDrawerResult =  Drawer()
                .withActivity(this)
                .withToolbar(toolbar)
                .withActionBarDrawerToggle(true)
                .withHeader(R.layout.drawer_header)
                .addDrawerItems(
                        PrimaryDrawerItem().withName(R.string.drawer_item_notification).withIcon(FontAwesome.Icon.faw_envelope).withBadge("1").withIdentifier(1).withIdentifier(1),
                        SecondaryDrawerItem().withName(R.string.drawer_item_help).withIcon(FontAwesome.Icon.faw_cog).withIdentifier(2),
                        SecondaryDrawerItem().withName(R.string.drawer_item_about).withIcon(FontAwesome.Icon.faw_user).withIdentifier(3),
                        DividerDrawerItem(),
                        SecondaryDrawerItem().withName(R.string.drawer_item_exit).withIcon(FontAwesome.Icon.faw_power_off).withIdentifier(4)
                ).withOnDrawerItemClickListener { parent, view, position, id, drawerItem ->

                    when(id) {

                        0.toLong() -> startActivity(Intent(this, NotificationListActivity::class.java))

                        1.toLong() -> startActivity(Intent(this, SettingsActivity::class.java))

                        2.toLong() -> startActivity(Intent(this, AboutActivity::class.java))

                        4.toLong() -> {
                            moveTaskToBack(true)
                            android.os.Process.killProcess(android.os.Process.myPid())
                            System.exit(1)
                        }
                    }

                }

                .build()

        // TODO = remove = Debug: for SQLite view
        Stetho.initializeWithDefaults(this)

        presenter.getAllHabits()
    }

    override fun onBackPressed() {

        if (mDrawerResult.isDrawerOpen) {
            mDrawerResult.closeDrawer()
        } else {
            super.onBackPressed()
        }
    }

    override fun showHabit(habitEntity: HabitEntity) {

        val habitCardFragment = HabitCardFragment()

        habitCardFragment.mHabitEntity = habitEntity

        supportFragmentManager.beginTransaction()
                .add(R.id.activity_home_content, habitCardFragment)
                .commit()
    }

    override fun showLoading() {

        findViewById<ProgressBar>(R.id.activity_home_progress_bar)?.visibility = View.VISIBLE
    }

    override fun hideLoading() {

        findViewById<ProgressBar>(R.id.activity_home_progress_bar)?.visibility = View.GONE
    }

    override fun showAddHabitDialog() {

        val promptsView = layoutInflater.inflate(R.layout.input_alert_add_habit, null)

        val alertDialogBuilder = AlertDialog.Builder(this)

        alertDialogBuilder.setView(promptsView)

        val habitNameEditText = promptsView.findViewById<EditText>(R.id.input_alert_add_habit_name)
        val timeCheckBox = promptsView.findViewById<CheckBox>(R.id.input_alert_add_habit_time)
        val moneyCheckBox = promptsView.findViewById<CheckBox>(R.id.input_alert_add_habit_money)
        val healthCheckBox = promptsView.findViewById<CheckBox>(R.id.input_alert_add_habit_health)

        alertDialogBuilder.setCancelable(false)

        alertDialogBuilder.setPositiveButton("Добавить") { _: DialogInterface, _: Int ->

            presenter.habitAddPositive (
                    habitNameEditText.text.toString(),
                    timeCheckBox.isChecked,
                    moneyCheckBox.isChecked,
                    healthCheckBox.isChecked
            )
        }

        alertDialogBuilder.setNegativeButton("Отмена") { dialog: DialogInterface, _: Int ->
            dialog.cancel()
        }

        val alertDialog = alertDialogBuilder.create()

        alertDialog.show()
    }


    @OnClick(R.id.activity_home_fab)
    fun addHabitClicked(view: View) {

       presenter.habitAddStart()
    }

    override fun getContext(): Context {

        return this@HomeActivity
    }
}
